### ISSUE TITLE
[Visual Studio] [C# 14] Private extension methods incorrectly marked as unused members

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#81657]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Visual-Studio-C-14-Private-extensio/11014236)._

---
[severity:It bothers me. A fix would be nice]
When using the new C# 14 extension method syntax, Visual Studio 18.0.2 marks private extension methods as "unused members" but only _if it is not called using the extension method signature_.
Surely this is the side-effect of some compiler lowering to old-style extension syntax, but the IDE suggestion is just wrong and would lead to a compiler error.

Example:

```csharp
public static class NewStyleExtensions
{
    extension(string input)
    {
        public T[] SplitToArray&lt;T&gt;(char separator) where T : IConvertible
            =&gt; [.. input.Split(separator).Select(ConvertTo&lt;T&gt;)];

        private T ConvertTo&lt;T&gt;() where T : IConvertible   // &lt;--- This method gets wrongfully marked as "unused member"
            =&gt; (T)Convert.ChangeType(input, typeof(T));
    }
}
```

Note that changing the call from the public method affects this behavior:
```csharp
// Alternative 1
// When called with the true extension method signature, the private method is now correctly marked as "used".
public T[] SplitToArray&lt;T&gt;(char separator) where T : IConvertible
    =&gt; [.. input.Split(separator).Select(x =&gt; x.ConvertTo&lt;T&gt;())];

// Alternative 2
// When called with the target as an input parameter, the private method is incorrectly marked as "unused" and a refactor is suggested to remove the lambda expression (to original example above).
public T[] SplitToArray&lt;T&gt;(char separator) where T : IConvertible
    =&gt; [.. input.Split(separator).Select(x =&gt; ConvertTo&lt;T&gt;(x))]; 
```

---
#### Benjamin Copass on 12/11/2025, 01:27 AM: 

<p>Another example / use case:</p>
<pre><code><span class="hljs-keyword">internal</span> <span class="hljs-keyword">static</span> <span class="hljs-keyword">class</span> <span class="hljs-title">PrivateExtensionPropertyTest</span>
{
    extension(<span class="hljs-keyword">object</span> o)
    {
        <span class="hljs-keyword">private</span> <span class="hljs-keyword">bool</span> Foo =&gt; o <span class="hljs-keyword">is</span> <span class="hljs-literal">null</span>; <span class="hljs-comment">// faded out in Visual Studio (VS thinks it's an unused member)</span>
    }

<span class="hljs-function"><span class="hljs-keyword">public</span> <span class="hljs-keyword">static</span> <span class="hljs-keyword">void</span> <span class="hljs-title">Test</span>(<span class="hljs-params"></span>)</span>
    {
        <span class="hljs-keyword">object</span> o = <span class="hljs-keyword">new</span> <span class="hljs-keyword">object</span>();
        <span class="hljs-keyword">bool</span> b = o.Foo; <span class="hljs-comment">// Foo is actually used</span>
    }
}
</code></pre>

#### Feedback Bot on 12/11/2025, 06:50 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

**Comment by deepakrathore33** [comment]
<img width="880" height="360" alt="Image" src="https://github.com/user-attachments/assets/6e035246-bb3e-4d93-b0f0-5ce6f796de04" />

**Comment by deepakrathore33** [comment]
Related to https://github.com/dotnet/roslyn/issues/80645

**Comment by jcouv** [comment]
Moved to Area-IDE as this seems driven by IDE0051
Also relates to https://github.com/dotnet/roslyn/issues/81213

I'm not sure if this is handled by the same IDE logic, but extension members and their implementation methods are also not tracked together in FAR and other refactorings: https://github.com/dotnet/roslyn/issues/81507

**Comment by jcouv** [comment]
Closing as duplicate of https://github.com/dotnet/roslyn/issues/80645
I'm working on a fix for 18.3
